File: android/app/src/main/java/com/scmessenger/android/ui/viewmodels/SettingsViewModel.kt
package com.scmessenger.android.ui.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.scmessenger.android.BuildConfig
import com.scmessenger.android.data.MeshRepository
import com.scmessenger.android.data.PreferencesRepository
import com.scmessenger.android.network.DiagnosticsReporter
import com.scmessenger.android.ui.diagnostics.DiagnosticsBundleFormatter
import com.scmessenger.android.ui.diagnostics.DiagnosticsBundleInput
import com.scmessenger.android.utils.Permissions
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import kotlinx.coroutines.delay
import timber.log.Timber
import javax.inject.Inject
import java.util.concurrent.atomic.AtomicLong
import kotlin.concurrent.Volatile

/**
 * ViewModel for the settings screen.
 *
 * Manages mesh settings, app preferences, and configuration.
 */
@HiltViewModel
class SettingsViewModel @Inject constructor(
    private val meshRepository: MeshRepository,
    private val preferencesRepository: PreferencesRepository,
    private val diagnosticsReporter: DiagnosticsReporter
) : ViewModel() {

    // Mesh settings
    private val _settings = MutableStateFlow<uniffi.api.MeshSettings?>(null)
    val settings: StateFlow<uniffi.api.MeshSettings?> = _settings.asStateFlow()

    // ANR FIX: Debouncing for settings updates to prevent UI thread spew
    // Use AtomicLong for thread-safe timestamp tracking without locks
    private val lastSettingUpdateNs = AtomicLong(0L)
    private val settingDebounceNs = 500_000_000L  // 500ms debounce

    // App preferences
    val autoStart = preferencesRepository.serviceAutoStart
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = false
        )

    val vpnMode = preferencesRepository.vpnModeEnabled
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = false
        )

    val themeMode = preferencesRepository.themeMode
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = PreferencesRepository.ThemeMode.SYSTEM
        )

    val notificationsEnabled = preferencesRepository.notificationsEnabled
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = true
        )

    val showPeerCount = preferencesRepository.showPeerCount
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = true
        )

    val autoAdjustEnabled = preferencesRepository.autoAdjustEnabled
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = true
        )

    val adjustmentProfile = preferencesRepository.manualAdjustmentProfile
        .map { name ->
            try {
                uniffi.api.AdjustmentProfile.valueOf(name)
            } catch (e: IllegalArgumentException) {
                uniffi.api.AdjustmentProfile.STANDARD
            }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = uniffi.api.AdjustmentProfile.STANDARD
        )

    // BLE Privacy (mirrors iOS BLE rotation settings)
    val bleRotationEnabled = preferencesRepository.bleRotationEnabled
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = true
        )

    val bleRotationIntervalSec = preferencesRepository.bleRotationIntervalSec
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = 900
        )

    // Loading state
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    val isSaving: StateFlow<Boolean> = _isLoading.asStateFlow()

    // Error state
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    // Identity
    private val _identityInfo = MutableStateFlow<uniffi.api.IdentityInfo?>(null)
    val identityInfo: StateFlow<uniffi.api.IdentityInfo?> = _identityInfo.asStateFlow()
    val hasIdentity: StateFlow<Boolean> = _identityInfo.map { it?.initialized == true }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), false)

    // Import result
    private val _importResult = MutableStateFlow<String?>(null)
    val importResult: StateFlow<String?> = _importResult.asStateFlow()

    // ANR FIX: Cached settings to avoid re-calculating on every composition
    private var cachedSettings: uniffi.api.MeshSettings? = null
    private var cachedIdentityInfo: uniffi.api.IdentityInfo? = null
    private var isCacheValid = false

    // ANR FIX: Guard identity emission — only emit when data actually changed.
    // Prevents infinite recomposition loops when getIdentityInfo() returns
    // structurally equal objects that would otherwise cascade through
    // StateFlow → hasIdentity → LaunchedEffect(hasIdentity) → loadIdentity().
    private fun emitIdentityInfo(info: uniffi.api.IdentityInfo?) {
        if (_identityInfo.value != info) {
            _identityInfo.value = info
            cachedIdentityInfo = info
        }
    }

    init {
        // ANR FIX (P0_ANDROID_017): Defer heavy initialization to background thread
        // Settings screen should appear within 500ms for good UX
        // Set defaults immediately so UI never shows missing sections
        _settings.value = uniffi.api.MeshSettings(
            relayEnabled = true,
            maxRelayBudget = 200u,
            batteryFloor = 20u,
            bleEnabled = true,
            wifiAwareEnabled = true,
            wifiDirectEnabled = true,
            internetEnabled = true,
            discoveryMode = uniffi.api.DiscoveryMode.NORMAL,
            onionRouting = false,
            coverTrafficEnabled = false,
            messagePaddingEnabled = false,
            timingObfuscationEnabled = false,
            notificationsEnabled = true,
            notifyDmEnabled = true,
            notifyDmRequestEnabled = true,
            notifyDmInForeground = false,
            notifyDmRequestInForeground = true,
            soundEnabled = true,
            badgeEnabled = true
        )

        viewModelScope.launch(kotlinx.coroutines.Dispatchers.IO) {
            try {
                _isLoading.value = true
                // Load settings first (depends on MeshRepository which may trigger service startup)
                loadSettingsInternal()
                // Then load identity
                loadIdentityInternal()
                // Mark cache as valid after initial load
                isCacheValid = true
            } catch (e: Exception) {
                Timber.e(e, "Failed to initialize SettingsViewModel")
            } finally {
                _isLoading.value = false
            }
        }
    }

    /**
     * Internal settings loader that runs on IO dispatcher.
     * ANR FIX: Uses cached settings when available to avoid redundant I/O.
     */
    private suspend fun loadSettingsInternal() {
        // ANR FIX: Return cached settings immediately if already loaded
        if (cachedSettings != null && isCacheValid) {
            _settings.value = cachedSettings
            return
        }
        try {
            val settings = meshRepository.loadSettings()
            cachedSettings = settings  // ANR FIX: Cache for future use
            _settings.value = settings
            Timber.d("Loaded mesh settings: $settings")
        } catch (e: Exception) {
            _error.value = "Failed to load settings: ${e.message}"
            Timber.e(e, "Failed to load mesh settings")
        }
    }

    /**
     * Internal identity loader that runs on IO dispatcher.
     * ANR FIX: Uses cached identity info when available to avoid redundant FFI calls.
     * Uses non-blocking identity access to avoid main thread blocking during service startup.
     */
    private suspend fun loadIdentityInternal() {
        // ANR FIX: Return cached identity if already loaded
        if (cachedIdentityInfo != null && isCacheValid) {
            emitIdentityInfo(cachedIdentityInfo)
            return
        }
        try {
            // Use non-blocking identity access to avoid blocking main thread
            // during service startup when identity might not be fully initialized yet
            val info = meshRepository.getIdentityInfoNonBlocking()
            // Defensive fallback: if Rust core returns blank nickname, use cached preference
            if (info != null && info.nickname.isNullOrBlank()) {
                val cached = preferencesRepository.identityNickname.first()
                if (!cached.isNullOrBlank()) {
                    Timber.w("Rust core nickname blank; using DataStore fallback: $cached")
                    // Push fallback to Rust Core to permanently fix the null state
                    meshRepository.setNickname(cached)
                    emitIdentityInfo(info.copy(nickname = cached))
                    return
                }
            }
            emitIdentityInfo(info)
        } catch (e: Exception) {
            Timber.e(e, "Failed to load identity")
        }
    }

    fun loadIdentity() {
        viewModelScope.launch(kotlinx.coroutines.Dispatchers.IO) {
            try {
                // ANR FIX: Return cached identity if already loaded — prevents
                // redundant FFI calls during recomposition cascades.
                if (cachedIdentityInfo != null && isCacheValid) {
                    emitIdentityInfo(cachedIdentityInfo)
                    return@launch
                }
                // Use non-blocking variant to avoid triggering service init
                // from the UI layer (prevents serviceState → loadIdentity loop).
                val info = meshRepository.getIdentityInfoNonBlocking()
                // Defensive fallback: if Rust core returns blank nickname, use cached preference
                if (info != null && info.nickname.isNullOrBlank()) {
                    val cached = preferencesRepository.identityNickname.first()
                    if (!cached.isNullOrBlank()) {
                        Timber.w("Rust core nickname blank; using DataStore fallback: $cached")
                        // Push fallback to Rust Core to permanently fix the null state
                        meshRepository.setNickname(cached)
                        emitIdentityInfo(info.copy(nickname = cached))
                        return@launch
                    }
                }
                emitIdentityInfo(info)
            } catch (e: Exception) {
                Timber.e(e, "Failed to load identity")
            }
        }
    }

    suspend fun getIdentityExportString(): String {
        // ANR FIX: Run on IO dispatcher to avoid blocking Main thread
        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            meshRepository.getIdentityExportString()
        }
    }

    /** P1_ANDROID_003: Import identity from a backup string. */
    fun importIdentityBackup(backup: String) {
        viewModelScope.launch(kotlinx.coroutines.Dispatchers.IO) {
            try {
                meshRepository.restoreIdentityFromBackup(backup)
                _importResult.value = "Identity restored successfully"
                loadIdentity() // Refresh UI after import
            } catch (e: Exception) {
                _importResult.value = "Failed to restore identity: ${e.message}"
                Timber.e(e, "Failed to restore identity from backup")
            }
        }
    }

    fun clearImportResult() {
        _importResult.value = null
    }

    fun updateNickname(name: String) {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                meshRepository.setNickname(name)
                loadIdentity() // Refresh to reflect change
            } catch (e: Exception) {
                _error.value = "Failed to update nickname: ${e.message}"
                Timber.e(e, "Failed to update nickname")
            }
        }
    }

    // ========================================================================
    // MESH SETTINGS
    // ========================================================================

    /**
     * Load mesh settings from repository.
     */
    fun loadSettings() {
        viewModelScope.launch {
            try {
                _isLoading.value = true
                _error.value = null

                val settings = meshRepository.loadSettings()
                _settings.value = settings

                Timber.d("Loaded mesh settings: $settings")
            } catch (e: Exception) {
                _error.value = "Failed to load settings: ${e.message}"
                Timber.e(e, "Failed to load mesh settings")
            } finally {
                _isLoading.value = false
            }
        }
    }

    /**
     * ANR FIX (P0_ANDROID_017): Debounced settings update.
     * Prevents rapid successive updates from causing excessive I/O and UI thread blocking.
     * Uses timestamp-based throttling with 500ms minimum interval between updates.
     */
    fun debouncedUpdateSettings(settings: uniffi.api.MeshSettings) {
        val nowNs = System.nanoTime()
        val lastUpdateNs = lastSettingUpdateNs.get()
        val timeSinceLastUpdateNs = nowNs - lastUpdateNs

        if (timeSinceLastUpdateNs < settingDebounceNs) {
            // Still within debounce window - log and skip
            Timber.d("Settings update throttled (${timeSinceLastUpdateNs / 1_000_000}ms since last)")
            return
        }

        // Try to atomically update the timestamp
        if (lastSettingUpdateNs.compareAndSet(lastUpdateNs, nowNs)) {
            // Success - we got the lock, proceed with update
            viewModelScope.launch {
                try {
                    // Validate settings
                    if (!meshRepository.validateSettings(settings)) {
                        _error.value = "Invalid settings configuration"
                        Timber.w("Invalid settings configuration")
                        return@launch
                    }

                    meshRepository.saveSettings(settings)
                    _settings.value = settings
                    Timber.i("Mesh settings saved (debounced)")
                } catch (e: Exception) {
                    _error.value = "Failed to save settings: ${e.message}"
                    Timber.e(e, "Failed to save mesh settings")
                }
            }
        } else {
            // Failed to atomically update - someone else beat us, skip this update
            Timber.d("Settings update skipped (concurrent update in progress)")
        }
    }

    /**
     * Update mesh settings - now redirects to debounced version.
     */
    fun updateSettings(settings: uniffi.api.MeshSettings) {
        debouncedUpdateSettings(settings)
    }

    /**
     * Update specific mesh setting fields - now uses debounced updates.
     */
    fun updateRelayEnabled(enabled: Boolean) {
        _settings.value?.let { current ->
            // Allow toggling - when disabled, stops ALL communication (bidirectional)
            debouncedUpdateSettings(current.copy(relayEnabled = enabled))
            // Apply transport settings when mesh participation changes
            meshRepository.applyTransportSettings(current.copy(relayEnabled = enabled))
        }
    }

    fun updateMaxRelayBudget(budget: UInt) {
        _settings.value?.let { current ->
            debouncedUpdateSettings(current.copy(maxRelayBudget = budget))
        }
    }

    fun updateBatteryFloor(floor: UByte) {
        _settings.value?.let { current ->
            debouncedUpdateSettings(current.copy(batteryFloor = floor))
        }
    }

    fun updateBleEnabled(enabled: Boolean) {
        _settings.value?.let { current ->
            debouncedUpdateSettings(current.copy(bleEnabled = enabled))
            // Apply transport settings to enable/disable BLE
            meshRepository.applyTransportSettings(current.copy(bleEnabled = enabled))
        }
    }

    fun updateWifiAwareEnabled(enabled: Boolean) {
        _settings.value?.let { current ->
            debouncedUpdateSettings(current.copy(wifiAwareEnabled = enabled))
            // Apply transport settings to enable/disable WiFi Aware
            meshRepository.applyTransportSettings(current.copy(wifiAwareEnabled = enabled))
        }
    }

    fun updateWifiDirectEnabled(enabled: Boolean) {
        _settings.value?.let { current ->
            debouncedUpdateSettings(current.copy(wifiDirectEnabled = enabled))
            // Apply transport settings to enable/disable WiFi Direct
            meshRepository.applyTransportSettings(current.copy(wifiDirectEnabled = enabled))
        }
    }

    fun updateInternetEnabled(enabled: Boolean) {
        _settings.value?.let { current ->
            debouncedUpdateSettings(current.copy(internetEnabled = enabled))
            // Apply transport settings to enable/disable the internet transport
            meshRepository.applyTransportSettings(current.copy(internetEnabled = enabled))
        }
    }

    fun updateDiscoveryMode(mode: uniffi.api.DiscoveryMode) {
        _settings.value?.let { current ->
            debouncedUpdateSettings(current.copy(discoveryMode = mode))
        }
    }

    /**
     * Reset mesh settings to factory defaults.
     */
    fun resetSettingsToDefault() {
        viewModelScope.launch(Dispatchers.IO) {
            try {
                _isLoading.value = true
                val defaults = meshRepository.getDefaultSettings()
                debouncedUpdateSettings(defaults)
                Timber.i("Mesh settings reset to defaults")
            } catch (e: Exception) {
                _error.value = "Failed to reset settings: ${e.message}"
                Timber.e(e, "Failed to reset mesh settings")
            } finally {
                _isLoading.value = false
            }
        }
    }



    // ========================================================================
    // APP PREFERENCES
    // ========================================================================

    fun setAutoStart(enabled: Boolean) {
        viewModelScope.launch {
            preferencesRepository.setServiceAutoStart(enabled)
        }
    }

    fun setVpnMode(enabled: Boolean) {
        viewModelScope.launch {
            preferencesRepository.setVpnMode(enabled)
        }
    }

    fun setThemeMode(mode: PreferencesRepository.ThemeMode) {
        viewModelScope.launch {
            preferencesRepository.setThemeMode(mode)
        }
    }

    fun setNotificationsEnabled(enabled: Boolean) {
        viewModelScope.launch {
            preferencesRepository.setNotificationsEnabled(enabled)
        }
    }

    fun setShowPeerCount(show: Boolean) {
        viewModelScope.launch {
            preferencesRepository.setShowPeerCount(show)
        }
    }

    // ========================================================================
    // POWER SETTINGS
    // ========================================================================

    fun setAutoAdjust(enabled: Boolean) {
        viewModelScope.launch {
            preferencesRepository.setAutoAdjustEnabled(enabled)
            if (enabled) {
                meshRepository.clearAdjustmentOverrides()
            }
        }
    }

    fun setManualProfile(profile: uniffi.api.AdjustmentProfile) {
        viewModelScope.launch {
            preferencesRepository.setManualAdjustmentProfile(profile.name)

            // Apply profile settings if manual
            // Note: In a real implementation, we might apply these to the engine directly
            // or the engine observes this preference.
            // For now, we just save the preference.
        }
    }

    fun overrideBleScanInterval(intervalMs: UInt) {
        meshRepository.overrideBleInterval(intervalMs)
    }

    fun overrideRelayMax(max: UInt) {
        meshRepository.overrideRelayMax(max)
    }

    fun clearAdjustmentOverrides() {
        meshRepository.clearAdjustmentOverrides()
    }

    // ========================================================================
    // UTILITIES
    // ========================================================================

    /**
     * Clear error state.
     */
    fun clearError() {
        _error.value = null
    }

    /**
     * Get ledger summary for diagnostics.
     */
    fun getLedgerSummary(): String {
        return meshRepository.getLedgerSummary()
    }

    fun getConnectionPathState(): uniffi.api.ConnectionPathState {
        return meshRepository.getConnectionPathState()
    }

    fun getNatStatus(): String {
        return meshRepository.getNatStatus()
    }

    /**
     * ANR FIX (P0_ANDROID_017): Export diagnostics asynchronously.
     * Original exportDiagnostics() was blocking on file I/O and Rust FFI calls.
     */
    suspend fun exportDiagnostics(): String {
        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            meshRepository.exportDiagnostics()
        }
    }

    fun getDiagnosticsLogPath(): String {
        return meshRepository.getDiagnosticsLogPath()
    }

    /**
     * ANR FIX (P0_ANDROID_017): Get diagnostics logs asynchronously.
     * Reading log files was blocking on Main thread.
     */
    suspend fun getDiagnosticsLogs(limit: Int = 500): String {
        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            meshRepository.getDiagnosticsLogs(limit)
        }
    }

    fun clearDiagnosticsLogs() {
        meshRepository.clearDiagnosticsLogs()
    }

    /** P0_ANDROID_007: Retry bootstrap with fallback strategy from diagnostics UI. */
    fun retryBootstrap() {
        viewModelScope.launch {
            try {
                meshRepository.bootstrapWithFallbackStrategy()
                Timber.i("Bootstrap retry initiated")
            } catch (e: Exception) {
                Timber.e(e, "Bootstrap retry failed")
            }
        }
    }

    /**
     * ANR FIX (P0_ANDROID_017): Build diagnostics bundle asynchronously.
     * All blocking I/O operations (exportDiagnostics, getPendingOutboxCount, getDiagnosticsLogs)
     * are moved to IO dispatcher to prevent UI thread blocking.
     */
    suspend fun buildTesterDiagnosticsBundle(): String {
        // Run the entire operation on IO dispatcher to avoid Main thread I/O
        return kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
            val missingPermissions = meshRepository.getMissingRuntimePermissions().map { permission ->
                "${Permissions.getPermissionName(permission)} ($permission)"
            }
            DiagnosticsBundleFormatter.format(
                DiagnosticsBundleInput(
                    generatedAtEpochMs = System.currentTimeMillis(),
                    appVersion = BuildConfig.VERSION_NAME,
                    serviceState = meshRepository.getServiceStateName(),
                    connectionPathState = meshRepository.getConnectionPathState().name,
                    natStatus = meshRepository.getNatStatus(),
                    discoveredPeers = meshRepository.getDiscoveredPeerCount(),
                    pendingOutbox = meshRepository.loadPendingOutbox().size,
                    missingPermissions = missingPermissions,
                    coreDiagnosticsJson = meshRepository.exportDiagnostics(),
                    recentLogs = meshRepository.getDiagnosticsLogs(limit = 1500)
                )
            )
        }
    }

    // ========================================================================
    // DIAGNOSTICS
    // ========================================================================

    /**
     * Get network diagnostics report from DiagnosticsReporter.
     */
    suspend fun getNetworkDiagnosticsReport(): DiagnosticsReporter.NetworkDiagnosticsReport? {
        return try {
            diagnosticsReporter.generateReport()
        } catch (e: Exception) {
            Timber.e(e, "Failed to generate diagnostics report")
            null
        }
    }



    // MARK: - Identity Helpers
    /**
     * Get contact count for info display.
     */
    fun getContactCount(): UInt {
        return meshRepository.getContactCount()
    }

    /**
     * Get message count for info display.
     */
    fun getMessageCount(): UInt {
        return meshRepository.getMessageCount()
    }

    // ========================================================================
    // BLE PRIVACY (mirrors iOS SettingsViewModel BLE rotation)
    // ========================================================================

    fun setBleRotationEnabled(enabled: Boolean) {
        viewModelScope.launch {
            preferencesRepository.setBleRotationEnabled(enabled)
        }
    }

    fun setBleRotationIntervalSec(intervalSec: Int) {
        viewModelScope.launch {
            preferencesRepository.setBleRotationIntervalSec(intervalSec)
        }
    }

    /**
     * Resets all application data and preferences.
     */
    fun resetAllData() {
        viewModelScope.launch {
            try {
                _isLoading.value = true

                // Clear app-level preferences
                preferencesRepository.clearAll()

                // Reset mesh data (identity, history, etc)
                meshRepository.resetAllData()

                Timber.i("Application reset complete")
            } catch (e: Exception) {
                _error.value = "Failed to reset application: ${e.message}"
                Timber.e(e, "Reset failed")
            } finally {
                _isLoading.value = false
            }
        }
    }
}

File: iOS/SCMessenger/SCMessenger/ViewModels/SettingsViewModel.swift
//
//  SettingsViewModel.swift
//  SCMessenger
//
//  ViewModel for settings management
//  Mirrors: android/.../ui/viewmodels/SettingsViewModel.kt
//

import Foundation
import os

@MainActor
@Observable
final class SettingsViewModel {
    private weak var repository: MeshRepository?

    var settings: MeshSettings?
    var isLoading = false
    var isSaving = false
    var error: String?
    var successMessage: String?
    private var lastSettingUpdateTime: Date = .distantPast
    private let settingDebounceInterval: TimeInterval = 0.5  // 500ms

    init(repository: MeshRepository) {
        self.repository = repository
        self.loadNickname()
    }

    private func debouncedUpdateSettings(_ update: @escaping () -> Void) {
        let now = Date()
        if now.timeIntervalSince(lastSettingUpdateTime) < settingDebounceInterval {
            os_log("Settings update throttled")
            return
        }
        lastSettingUpdateTime = now
        update()
    }

    // MARK: - Identity

    var nickname: String = ""

    func loadNickname() {
        if let name = repository?.getNickname() {
            self.nickname = name
        } else {
            // A relay-only installation has no identity yet.  Showing a fake
            // default name makes Settings look initialized and can accidentally
            // create an identity with an unintended nickname.
            self.nickname = ""
        }
    }

    func updateNickname(_ name: String) {
        Task { @MainActor in
            do {
                try await repository?.setNickname(name)
                self.nickname = name
                successMessage = "Nickname updated"
            } catch {
                self.error = "Failed to update nickname: \(error.localizedDescription)"
            }
        }
    }

    func getIdentityExportString() async -> String {
        return await repository?.getIdentityExportString() ?? "{}"
    }

    // MARK: - Identity Backup (passphrase-encrypted)

    var backupExportResult: Result<String, Error>?
    var isExportingBackup = false

    /// Export a passphrase-encrypted identity backup (identity key + ratchet
    /// sessions + contacts). Distinct from `getIdentityExportString()`, which
    /// exports the public identity card with no encryption.
    ///
    /// The generated UniFFI surface follows this target's MainActor isolation.
    /// Yield first so the progress state is rendered before the synchronous
    /// encryption call begins.
    func exportIdentityBackup(passphrase: String) {
        guard let core = repository?.ironCore else {
            backupExportResult = .failure(MeshError.notInitialized("IronCore not initialized"))
            return
        }
        isExportingBackup = true
        Task { @MainActor [weak self] in
            await Task.yield()
            let result: Result<String, Error>
            do {
                result = .success(try core.exportIdentityBackup(passphrase: passphrase))
            } catch {
                result = .failure(error)
            }
            self?.finishBackupExport(result)
        }
    }

    private func finishBackupExport(_ result: Result<String, Error>) {
        isExportingBackup = false
        backupExportResult = result
    }

    func clearBackupExportResult() {
        backupExportResult = nil
    }

    var backupImportResult: Result<Void, Error>?
    var isImportingBackup = false

    /// Import an identity backup using a user-supplied passphrase. This mirrors
    /// `exportIdentityBackup`'s progress yield and completes repository-level
    /// identity publication after a successful import.
    func importIdentityBackup(backup: String, passphrase: String) {
        guard let core = repository?.ironCore else {
            backupImportResult = .failure(MeshError.notInitialized("IronCore not initialized"))
            return
        }
        isImportingBackup = true
        Task { @MainActor [weak self] in
            await Task.yield()
            let result: Result<Void, Error>
            do {
                try core.importIdentityBackup(backup: backup, passphrase: passphrase)
                if core.getIdentityInfo().initialized {
                    core.grantConsent()
                }
                result = .success(())
            } catch {
                result = .failure(error)
            }
            self?.finishBackupImport(result)
        }
    }

    private func finishBackupImport(_ result: Result<Void, Error>) {
        if case .success = result {
            repository?.finalizeImportedIdentity()
            loadNickname()
        }
        isImportingBackup = false
        switch result {
        case .success:
            successMessage = "Identity restored successfully"
        case .failure(let error):
            self.error = "Failed to restore identity: \(error.localizedDescription)"
        }
        backupImportResult = result
    }

    func clearBackupImportResult() {
        backupImportResult = nil
    }

    // MARK: - Settings Lifecycle

    func loadSettings() {
        isLoading = true
        do {
            settings = try repository?.loadSettings()
            error = nil
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func saveSettings() {
        guard let settings = settings else { return }

        isSaving = true
        do {
            try repository?.saveSettings(settings)
            successMessage = "Settings saved"
            error = nil
        } catch {
            self.error = error.localizedDescription
        }
        isSaving = false
    }

    // MARK: - Relay

    func toggleRelay() {
        guard var currentSettings = settings else { return }
        currentSettings.relayEnabled = !currentSettings.relayEnabled
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    // MARK: - Transport Toggles (mirrors Android MeshSettingsScreen)

    func updateBleEnabled(_ enabled: Bool) {
        guard var currentSettings = settings else { return }
        currentSettings.bleEnabled = enabled
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    func updateWifiAwareEnabled(_ enabled: Bool) {
        guard var currentSettings = settings else { return }
        currentSettings.wifiAwareEnabled = enabled
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    func updateWifiDirectEnabled(_ enabled: Bool) {
        guard var currentSettings = settings else { return }
        currentSettings.wifiDirectEnabled = enabled
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    func updateInternetEnabled(_ enabled: Bool) {
        guard var currentSettings = settings else { return }
        currentSettings.internetEnabled = enabled
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    // MARK: - Discovery & Mesh Settings

    func updateDiscoveryMode(_ mode: DiscoveryMode) {
        guard var currentSettings = settings else { return }
        currentSettings.discoveryMode = mode
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    func updateBatteryFloor(_ floor: UInt8) {
        guard var currentSettings = settings else { return }
        currentSettings.batteryFloor = floor
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    // MARK: - Privacy / Onion Routing

    func updateOnionRouting(_ enabled: Bool) {
        guard var currentSettings = settings else { return }
        currentSettings.onionRouting = enabled
        settings = currentSettings
        debouncedUpdateSettings { self.saveSettings() }
    }

    // MARK: - BLE Privacy

    var isBleRotationEnabled: Bool {
        return repository?.blePrivacyEnabled ?? true
    }

    var bleRotationInterval: TimeInterval {
        return repository?.blePrivacyInterval ?? 900
    }

    func toggleBleRotation(enabled: Bool) {
        repository?.blePrivacyEnabled = enabled
    }

    func updateBleRotationInterval(_ interval: TimeInterval) {
        repository?.blePrivacyInterval = interval
    }

    // MARK: - App Preferences (mirrors Android PreferencesRepository)

    var isNotificationsEnabled: Bool {
        get { settings?.notificationsEnabled ?? true }
        set {
            guard var currentSettings = settings else { return }
            currentSettings.notificationsEnabled = newValue
            settings = currentSettings
            debouncedUpdateSettings { self.saveSettings() }
            if newValue {
                Task {
                    _ = await NotificationManager.shared.requestPermissionIfNeeded()
                }
            }
        }
    }

    var notifyDmEnabled: Bool {
        get { settings?.notifyDmEnabled ?? true }
        set {
            guard var currentSettings = settings else { return }
            currentSettings.notifyDmEnabled = newValue
            settings = currentSettings
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    var notifyDmRequestEnabled: Bool {
        get { settings?.notifyDmRequestEnabled ?? true }
        set {
            guard var currentSettings = settings else { return }
            currentSettings.notifyDmRequestEnabled = newValue
            settings = currentSettings
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    var notifyDmInForeground: Bool {
        get { settings?.notifyDmInForeground ?? false }
        set {
            guard var currentSettings = settings else { return }
            currentSettings.notifyDmInForeground = newValue
            settings = currentSettings
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    var notifyDmRequestInForeground: Bool {
        get { settings?.notifyDmRequestInForeground ?? true }
        set {
            guard var currentSettings = settings else { return }
            currentSettings.notifyDmRequestInForeground = newValue
            settings = currentSettings
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    var soundEnabled: Bool {
        get { settings?.soundEnabled ?? true }
        set {
            guard var currentSettings = settings else { return }
            currentSettings.soundEnabled = newValue
            settings = currentSettings
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    var badgeEnabled: Bool {
        get { settings?.badgeEnabled ?? true }
        set {
            guard var currentSettings = settings else { return }
            currentSettings.badgeEnabled = newValue
            settings = currentSettings
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    // MARK: - Privacy Feature Preferences
    //
    // These preferences persist via UserDefaults and are forwarded to the Rust
    // core privacy module (cover traffic, padding, timing obfuscation) via
    // MeshSettings once the UniFFI surface exposes per-feature toggles.
    // Until then the stored value is the source of truth consulted by the UI.

    var isCoverTrafficEnabled: Bool {
        get { UserDefaults.standard.object(forKey: "privacy_cover_traffic") as? Bool ?? false }
        set {
            UserDefaults.standard.set(newValue, forKey: "privacy_cover_traffic")
            guard var s = settings else { return }
            s.coverTrafficEnabled = newValue
            settings = s
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    var isMessagePaddingEnabled: Bool {
        get { UserDefaults.standard.object(forKey: "privacy_message_padding") as? Bool ?? false }
        set {
            UserDefaults.standard.set(newValue, forKey: "privacy_message_padding")
            guard var s = settings else { return }
            s.messagePaddingEnabled = newValue
            settings = s
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    var isTimingObfuscationEnabled: Bool {
        get { UserDefaults.standard.object(forKey: "privacy_timing_obfuscation") as? Bool ?? false }
        set {
            UserDefaults.standard.set(newValue, forKey: "privacy_timing_obfuscation")
            guard var s = settings else { return }
            s.timingObfuscationEnabled = newValue
            settings = s
            debouncedUpdateSettings { self.saveSettings() }
        }
    }

    // MARK: - Service Control

    var isServiceRunning: Bool {
        return repository?.serviceState == .running
    }

    func startService() {
        repository?.start()
    }

    func stopService() {
        repository?.stopMeshService()
    }

    // MARK: - Info Counts

    func getContactCount() -> UInt32 {
        return (try? repository?.getContactCount()) ?? 0
    }

    func getMessageCount() -> UInt32 {
        return (try? repository?.getMessageCount()) ?? 0
    }

    func getConnectionPathState() -> ConnectionPathState {
        return repository?.getConnectionPathState() ?? .disconnected
    }

    func getNatStatus() -> String {
        return repository?.getNatStatus() ?? "unknown"
    }

    func exportDiagnostics() -> String {
        return repository?.exportDiagnostics() ?? "{}"
    }

    // MARK: - Auto-Adjust (mirrors Android PowerSettingsScreen)

    var isAutoAdjustEnabled: Bool {
        get { repository?.isAutoAdjustEnabled ?? (UserDefaults.standard.object(forKey: "auto_adjust_enabled") as? Bool ?? true) }
        set { repository?.setAutoAdjustEnabled(newValue) }
    }

    // MARK: - Factory Reset (mirrors Android resetAllData)

    /// Wipe all local data and return the app to first-run state.
    func resetAllData() async {
        await repository?.resetAllData()
        nickname = ""
        successMessage = nil
        error = nil
    }
}
